### PR TITLE
Add URI and text filters

### LINKED ISSUE #406: Implement filtering options for Text types.
## Problem
We need to ensure that records that include columns of all Text types support the following filters via API: 
- contains {x}
- is {x}
- is not {x}
- starts with {x}
- ends with {x}
- does not contain {x}
- does not start with {x}
- does not end with {x}
- is empty
- is not empty

This involves:
- Implementing the filters in the backend
- Updating the `/api/v0/databases/<id>/types/` endpoint to store available filters on this type
  - Filter information should include the number of parameters needing to be passed in (e.g. `contains` needs 1 parameter, `is empty` needs 0)

## Additional context
- We're using our fork of `sqlalchemy-filters` to provide filtering. See: https://github.com/centerofci/sqlalchemy-filters.
- This issue is blocked by implementation of Text types: 
  - #404 
  - #405 
  - #407
- #557 provides some context on storing filters.

### LINKED ISSUE #413: Implement filtering options for the URI type
## Problem
We need to ensure that records that include columns of the custom URI type support the following filters via API: 
- contains {x}
- does not contain {x}
- is {x}
- is not {x}
- is empty
- is not empty
- **host contains {x}**
- **host does not contain {x}**
- **scheme is [select from list of choices of valid schemes]**
- **scheme is not [select from list of choices of valid schemes]**

The ones in bold are unique to the URI type.

This involves:
- Implementing the filters in the backend
- Updating the `/api/v0/databases/<id>/types/` endpoint to store available filters on this type
  - Filter information should include the number of parameters needing to be passed in (e.g. `is` needs 1 parameter, `is empty` needs 0)
  - Filter information should also include the list of valid choices for the filter (e.g. scheme filters)

## Additional context
- We're using our fork of `sqlalchemy-filters` to provide filtering. See: https://github.com/centerofci/sqlalchemy-filters.
- Blocked by #412 
- #557 provides some context on storing filters.

### PR DESCRIPTION
**Pull request by dmos62**
Fixes #413, Fixes #406

Implements following filters:
- x contains y,
- x contains (case insensitive) y,
- x starts with (case insensitive) y,
- x URI authority contains y,
- x URI scheme equals y.

Other filters required in #413 and #406 have been implemented in previous PRs.

**Technical details**

The new filters look like this on the `filters` endpoint:

```json
    {
        "id": "contains",
        "name": "contains",
        "parameters": [
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            },
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            }
        ]
    },
    {
        "id": "starts_with_case_insensitive",
        "name": "starts with (case insensitive)",
        "parameters": [
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            },
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            }
        ]
    },
    {
        "id": "uri_authority_contains",
        "name": "URI authority contains",
        "parameters": [
            {
                "ui_types": [
                    "uri"
                ]
            },
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            }
        ]
    },
    {
        "id": "uri_scheme_equals",
        "name": "URI scheme equals",
        "parameters": [
            {
                "ui_types": [
                    "uri"
                ]
            },
            {
                "ui_types": [
                    "text",
                    "uri"
                ]
            }
        ]
    }
``` 

Notice that the Mathesar type `uri` is string-like, just like `text`, so it's allowed wherever `text` is allowed (see `contains` parameters' types for an example). In contrast, `uri_authority_contains` first parameter can only be `uri`. Same for `uri_scheme_equals`.

Tests for new filters have been added.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by dmos62** [comment]
Notice that in the issue #413, the `x URI scheme equals y` filter is specified to allow the user to choose from a list of URI schemes. This PR *does not* implement that.

It could be done somewhat easily. Just add to the relevant parameter a hint like `hint.suggested_values(("http","https","ftp","ftps",...))`. Or, give the parameter a hint specifying that it must be either just a string-like or a uri-scheme: then the frontend would know what kinds of suggestions would be appropriate there. This would require putting some `and`-`or` logic into the hint system though, which will be necessary down the road, but maybe not so much now.

At the same time, I think that scheme suggestions is not an important feature right now, so I propose we create an issue for it and put it off.

**Comment by dmos62** [comment]
Note that in this PR the filters endpoint response includes some filters multiple times. This is fixed in https://github.com/centerofci/mathesar/pull/1090.

File: db/functions/redundant.py
Comment (kgodey) [inline db/functions/redundant.py:14]: I find this name confusing. Perhaps something like `DBFunctionCombination` would be clearer?

File: db/functions/redundant.py
Comment (kgodey) [inline db/functions/redundant.py:73]: If this name is meant to be used in the frontend, please remove ` (case insensitive)` from the `name` parameter.

**Review by kgodey — CHANGES_REQUESTED** [review]
A couple of changes requested in individual comments below. I'd also like @mathemancer to take a look at this before merge since it touches the URI type.

> Notice that in the issue #413, the `x URI scheme equals y` filter is specified to allow the user to choose from a list of URI schemes. This PR *does not* implement that.
> 
> At the same time, I think that scheme suggestions is not an important feature right now, so I propose we create an issue for it and put it off.

I think we should do this, it is important for making the product friendly to non-technical users, who may not know what a URI scheme is.

> It could be done somewhat easily. Just add to the relevant parameter a hint like `hint.suggested_values(("http","https","ftp","ftps",...))`.
>
> This would require putting some `and`-`or` logic into the hint system though, which will be necessary down the road, but maybe not so much now.

I'm not sure why this would need "`and`-`or` logic",  can you elaborate?

File: db/functions/redundant.py
Comment (kgodey) [inline db/functions/redundant.py:92]: Case insensitive comment from above applies here as well.

File: db/functions/redundant.py
Comment (dmos62) [reply db/functions/redundant.py:73]: It is. If the suffix is removed, it will have the same display name as a regular (case-sensitive) `starts with` filter: a user won't be able to distinguish the two.

File: db/functions/redundant.py
Comment (dmos62) [reply db/functions/redundant.py:14]: I think DBFunctionCombination can be even more confusing, since a regular DBFunction instance can contain other DBFunction instances as parameters, and thus could be called a combination or supporting combination. I like redundant, because it highlights that it's made of DBFunctions that could be used without this particular DBFunction. I'm open to other names. Maybe `SecondaryDBFunction`?

**Comment by dmos62** [comment]
>> It could be done somewhat easily. Just add to the relevant parameter a hint like `hint.suggested_values(("http","https","ftp","ftps",...))`.
>>
>> This would require putting some `and`-`or` logic into the hint system though, which will be necessary down the road, but maybe not so much now.

>I'm not sure why this would need "`and`-`or` logic",  can you elaborate?

I wasn't clear enough. These are the options I mentioned:

1. > It could be done somewhat easily. Just add to the relevant parameter a hint like hint.suggested_values(("http","https","ftp","ftps",...)).
2. > Or, give the parameter a hint specifying that it must be either just a string-like or a uri-scheme: then the frontend would know what kinds of suggestions would be appropriate there. This would require putting some and-or logic into the hint system though, which will be necessary down the road, but maybe not so much now.

If this is wanted for alpha, I'll hack something together into a dedicated PR.

**Comment by dmos62** [comment]
@kgodey @mathemancer note that the changes in the `db/types/uri.py` are just me adding URI-specific DBFunctions to that namespace. I don't make changes to the exiting URI type code.

File: db/functions/redundant.py
Comment (kgodey) [reply db/functions/redundant.py:14]: I like `DBFunctionPackage`.

File: db/functions/redundant.py
Comment (kgodey) [reply db/functions/redundant.py:73]: There shouldn't be a case-sensitive "starts with" in the frontend, only the case insensitive version. 

We don't want to overwhelm users with a lot of similar filtering options, it will make finding the filter they want harder. And non-technical users may not know what "case sensitive" refers to.

**Comment by kgodey** [comment]
> If this is wanted for alpha, I'll hack something together into a dedicated PR.

Okay, thanks, please do. Option 1 seems easiest.

**Comment by dmos62** [comment]
Ready for review.

I've removed case-sensitive string filters from the filters endpoint.

Other requested changes will be submitted as dedicated PR/s. Those changes are:
  - rename RedundantDBFunction to something better;
  - have appropriate filters suggest popular/standard URI schemes
    - https://github.com/centerofci/mathesar/issues/1097

**Review by kgodey — APPROVED** [review]
Looks good to me, I'll leave it to @mathemancer to review/merge.

File: db/functions/redundant.py
Comment (mathemancer) [inline db/functions/redundant.py:25]: For my own education:  Should implementations of this `unpack` method recurse "in place"?  I.e., if I happen to have a redundant (or whatever we decide for a term) function that's composed of other redundant functions, should the `unpack` method call the unpack methods of those functions as well, or should that be left to the caller?  I'm not sure since you'd want to avoid having the recursion in multiple places, and `_db_function_to_sa_expression` is already recursive.

File: db/types/base.py
Comment (mathemancer) [inline db/types/base.py:159]: This sort of info is also in the `cast.py` file.  Did you consider trying to unify these with those definitions?  I ask because it makes me nervous to have this defined in more than one place.

**Review by mathemancer — CHANGES_REQUESTED** [review]
I mostly just had questions; the only real request is to try to unify the definitions of things like "string-like" "number-like", etc. between the definitions in the `db.types.operations.cast` module and the hint building logic.

File: db/functions/base.py
Comment (dmos62) [reply db/functions/base.py:233]: I find the shorthand tuple syntax (`(x,)`) to be errorprone. Just yesterday I ran into an annoying bug where I had forgotten the trailing comma and what was supposed to be a single-element tuple ended up being just the element. So I try to do the more awkward, but safer `tuple([x])`. This is not a very strong preference. After getting bit by this a few times you learn to append a comma to everything, I guess.

File: db/functions/base.py
Comment (dmos62) [reply db/functions/base.py:80]: I wanted to use 1-based indexing, but changed my mind. Everything in Python is 0-based, so I decided to go with the flow.

File: db/functions/redundant.py
Comment (dmos62) [reply db/functions/redundant.py:25]: I think it's desirable not to have to do manual recursion inside `unpack`. I don't immediately recall if I implemented full recursion yet. I think I did, since I was concerned about infinite loops.

File: db/types/base.py
Comment (dmos62) [reply db/types/base.py:159]: I didn't notice the duplication. I'll investigate.

**Comment by dmos62** [comment]
@mathemancer could we get this merged? It's the base for a few other PRs. I'll open an issue for the duplicated logic across `db.types.operations.cast` and `db.types.base` modules and see about fixing that in a new PR.

**Review by mathemancer — APPROVED** [review]
Approving, since I agree we should probably get this merged to unblock downstream PRs.  Please don't forget to create the issue about the duplicated definitions of type tags (string-like, number-like, etc.)

**Comment by dmos62** [comment]
@mathemancer the issue for that is here: https://github.com/centerofci/mathesar/issues/1100

**Pull request merged**
